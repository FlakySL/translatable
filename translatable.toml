
locales_path = "./translatable/tests/environments/everything_valid/translations/"
