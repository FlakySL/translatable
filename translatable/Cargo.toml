[package]
name = "translatable"
description = "A robust internationalization solution for Rust featuring compile-time validation, ISO 639-1 compliance, and TOML-based translation management. "
repository = "https://github.com/FlakySL/translatable.rs"
license = "MIT OR Apache-2.0"
readme = "../README.md"
version = "0.1.0"
edition = "2024"
authors = ["Esteve Autet <esteve@memw.es>", "chikof"]
tags = ["i18n", "translations", "idioms", "languages", "internazionalization"]

[dependencies]
thiserror = "2.0.12"
translatable_proc = { path = "../translatable_proc" }

[dev-dependencies]
trybuild = "1.0.104"
