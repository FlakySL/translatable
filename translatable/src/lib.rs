use thiserror::Error;

// re export the macro in the main crate.
pub use translatable_proc::translation;

/// This error is used on results for the
/// translation procedural macro, the macro
/// will return a Result<Option<&'static str>, Error>,
/// when there is a dynamic expression to resolve.
///
/// For example, if the language is a dynamic expression
/// meaning it's not a literal &'static str, and it evaluates
/// on runtime, if the runtime evaluation is invalid because
/// the language does not match the ISO 639-1 specification
/// or something else, the translation macro will return an
/// Error::InvalidLanguage.
///
/// For more information on the possible errors read each
/// enum branch documentation.
#[derive(Error, Debug)]
pub enum Error {
    #[error("The language '{0}' is invalid.")]
    InvalidLanguage(String),

    #[error("The langauge '{0}' is not available for the path '{1}'")]
    LanguageNotAvailable(String, String)
}

/// This module is for internal usage, it's members
/// are not documented, and there is no support on
/// using it.
#[doc(hidden)]
pub mod internal {
    use std::collections::HashMap;

    #[doc(hidden)]
    pub enum NestingType {
        Object(HashMap<String, NestingType>),
        Translation(HashMap<String, String>)
    }

    #[doc(hidden)]
    impl NestingType {
        pub fn get_path(&self, path: Vec<&str>) -> Option<&HashMap<String, String>> {
            match self {
                Self::Object(nested) => {
                    let (first, rest) = path.split_first()?;

                    nested
                        .get(*first)
                        .and_then(|n| n.get_path(rest.to_vec()))
                },

                Self::Translation(translation) => {
                    if path.is_empty() {
                        return Some(translation)
                    }

                    None
                }
            }
        }
    }
}
