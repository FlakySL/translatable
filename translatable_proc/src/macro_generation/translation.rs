//! [`translation!()`] macro output module.
//!
//! This module contains the required for
//! the generation of the [`translation!()`] macro tokens
//! with intrinsics from [`macro_input::translation`].
//!
//! The macro is separated by 4 functions, the main
//! function [`translation_macro`] which is exported
//! handles the context and conditionally calls the
//! other 3 functions based on the macro intrinsics.
//!
//! The language is really only used statically
//! if the path and language are both static, otherwise
//! static path is only validated but not used further
//! in compile time.
//!
//! [`translation!()`]: crate::translation
//! [`macro_input::translation`]: super::super::macro_input::translation

use proc_macro2::TokenStream as TokenStream2;
use quote::{ToTokens, quote};
use thiserror::Error;
use translatable_shared::{handle_macro_result, inline_quote};
use translatable_shared::macros::collections::{map_to_tokens, map_transform_to_tokens};
use translatable_shared::misc::language::Language;
use translatable_shared::misc::templating::FormatString;
use translatable_shared::translations::collection::TranslationNodeCollection;
use translatable_shared::translations::node::TranslationObject;

use crate::data::config::load_config;
use crate::data::translations::load_translations;
use crate::macro_input::translation::TranslationMacroArgs;
use crate::macro_input::utils::input_type::InputType;
use crate::macro_input::utils::translation_path::TranslationPath;

/// Macro compile-time translation resolution error.
///
/// Represents errors that can occur while compiling the [`translation!()`]
/// macro. This includes cases where a translation path cannot be found or
/// a language variant is unavailable at the specified path.
///
/// These errors are reported at compile-time by `rust-analyzer`
/// for immediate feedback while invoking the [`translation!()`] macro.
///
/// [`translation!()`]: crate::translation
#[derive(Error, Debug)]
enum MacroCompileError {
    /// The requested translation path could not be found.
    ///
    /// **Parameters**
    /// * `0` — The translation path, displayed in `::` notation.
    #[error("The path '{0}' could not be found")]
    PathNotFound(String),

    /// The requested language is not available for the provided translation
    /// path.
    ///
    /// **Parameters**
    /// * `0` — The requested `Language`.
    /// * `1` — The translation path where the language was expected.
    #[error("The language '{0:?}' ('{0:#}') is not available for the path '{1}'")]
    LanguageNotAvailable(Language, String),

    /// The fallback language is not available for the provided translation
    /// path.
    ///
    /// **Parameters**
    /// * `0` - The translation path where the language was expected.
    #[error("The configured fallback language is not available for this '{0}'.")]
    FallbackNotAvailable(String),
}

/// Local macro generation context.
///
/// Macro generation context used for concern separation
/// on generating the [`translation!()`] macro generation separatedly
/// in multiple functions.
///
/// [`translation!()`]: crate::translation
struct GenerationContext {
    /// The translation fallback language
    /// whether it's available or not.
    fallback_language: Option<Language>,

    /// A reference to the translation nodes
    /// loaded from the file system.
    translations: &'static TranslationNodeCollection,

    /// A [`TokenStream2`] pointing to the
    /// same address as the pre-processed
    /// template replacements that must
    /// be dynamically applied.
    template_replacements: TokenStream2,
}

/// [`TranslationObject`] static obtention helper.
///
/// Obtain a specific translation object in compile
/// time using macro intrinsics from a [`TranslationNodeCollection`]
/// and a [`TranslationPath`], converting the possible error
/// into the corresponding [`MacroCompileError`].
///
/// **Arguments**
/// * `translations` - The collection of files where to find the translation
///   object.
/// * `path` - The path to look for inside the collection.
///
/// **Returns**
/// A [`Result`] containing the translation object or a
/// [`MacroCompileError::PathNotFound`] error.
fn get_translation_object<'r>(
    translations: &'r TranslationNodeCollection,
    path: &TranslationPath,
) -> Result<&'r TranslationObject, MacroCompileError> {
    translations
        .find_path(path)
        .ok_or_else(|| MacroCompileError::PathNotFound(path.static_display()))
}

/// [`TranslationObject`] fallback helper.
///
/// Obtains the corresponding fallback translation
/// for a [`TranslationObject`], converting the possible
/// error to the corresponding [`MacroCompileError`].
///
/// **Arguments**
/// * `original_path` - The original path where the translation was found.
/// * `translation` - The translation object for where to find the fallback translation.
/// * `fallback_language` - The fallback language to find the translation.
///
/// **Returns**
/// [`MacroCompileError::FallbackNotAvailable`] if there is a fallback
/// but is not available in the translation otherwise [`Ok`] whether there
/// was a fallback language specified or not.
fn get_fallback_translation<'r>(
    original_path: &TranslationPath,
    translation: &'r TranslationObject,
    fallback_language: Option<Language>
) -> Result<Option<&'r FormatString>, MacroCompileError> {
    fallback_language
        .map(|lang| translation
            .get(&lang)
            .ok_or_else(|| MacroCompileError::FallbackNotAvailable(original_path.static_display())))
        .transpose()
}

/// Fully static arguments generation.
///
/// Concern separation for [`translation_macro`].
/// Generates the macro output taking all static arguments.
///
/// **Arguments**
/// * `ctx` - The macro generation context
/// * `language` - The static langauge argument
/// * `path` - The static path argument
///
/// **Returns**
/// Tokens to be directly returned for all static generation
#[inline(always)]
fn all_static(ctx: &GenerationContext, language: Language, path: &TranslationPath) -> TokenStream2 {
    let translation_object = handle_macro_result!(get_translation_object(ctx.translations, path));
    let fallback_translation = handle_macro_result!(get_fallback_translation(path, translation_object, ctx.fallback_language));

    let translation = handle_macro_result!(
        translation_object
            .get(&language)
            .or(fallback_translation)
            .ok_or_else(|| MacroCompileError::LanguageNotAvailable(
                language,
                path.static_display()
            ))
    );

    inline_quote! {{
        #translation
            .replace_with(&#{ctx.template_replacements})
    }}
}

/// Path static generation.
#[inline(always)]
fn path_static(
    ctx: &GenerationContext,
    lang: TokenStream2,
    path: &TranslationPath,
) -> TokenStream2 {
    let translation_object = handle_macro_result!(get_translation_object(ctx.translations, path));
    let fallback_translation = handle_macro_result!(get_fallback_translation(path, translation_object, ctx.fallback_language));

    inline_quote! {
        #{map_to_tokens(translation_object)}
    }
}

#[inline(always)]
fn all_dynamic(ctx: &GenerationContext, lang: TokenStream2, path: TokenStream2) -> TokenStream2 {
    quote!{}
}

/// [`translation!()`] macro output generation.
///
/// Expands into code that resolves a translation string based on the input
/// language and translation path, performing placeholder substitutions
/// if applicable.
///
/// If the language and path are fully static, the translation will be resolved
/// during macro expansion. Otherwise, the generated code will include runtime
/// resolution logic.
///
/// If the path or language is invalid at compile time, an appropriate
/// `MacroCompileError` will be reported.
///
/// **Arguments**
/// * `input` — Structured arguments defining the translation path, language,
/// and any placeholder replacements obtained from [`macro_input::translation`].
///
/// **Returns**
/// Generated `TokenStream2` representing the resolved translation string or
/// runtime lookup logic.
///
/// [`macro_input::translation`]: super::super::macro_input::translation
/// [`translation!()`]: crate::translation
pub fn translation_macro(input: TranslationMacroArgs) -> TokenStream2 {
    // TODO: separate concerns

    let config = handle_macro_result!(load_config());
    let translations = handle_macro_result!(load_translations());

    let ctx = GenerationContext {
        fallback_language: config.fallback_language(),
        translations,
        template_replacements: map_transform_to_tokens(
            input.replacements(),
            |key, value| quote! { (stringify!(#key).to_string(), #value.to_string()) },
        ),
    };

    todo!();

    if let InputType::Static(language) = input.language() {
        if let InputType::Static(path) = input.path() {
            return all_static(&ctx, language.clone(), path);
        }
    }

    let language = match input.language() {
        InputType::Static(language) => language.to_token_stream(),

        InputType::Dynamic(language) => quote! {
            translatable::shared::misc::language::Language::from(#language)
        },
    };

    let translation_object = match input.path() {
        InputType::Static(path) => {
            let path_segments = path.segments();
            let static_path_display = path_segments.join("::");

            let translation_object = handle_macro_result!(
                translations
                    .find_path(path_segments)
                    .ok_or_else(|| MacroCompileError::PathNotFound(static_path_display.clone()))
            );

            let translations_tokens = map_to_tokens(translation_object);

            if let Some(language_fallback) = config.fallback_language() {
                let translation_fallback = handle_macro_result!(
                    translation_object
                        .get(&language_fallback)
                        .ok_or_else(|| MacroCompileError::FallbackNotAvailable(
                            static_path_display
                        ))
                );

                return quote! {{
                    #translations_tokens
                        .get(&#language)
                        .unwrap_or_else(|| #translation_fallback)
//                        .replace_with(&#template_replacements)
                }};
            }

            quote! {
                #[doc(hidden)]
                let path: Vec<_> = vec![#(#path_segments.to_string()),*];

                #translations_tokens
            }
        },

        InputType::Dynamic(path) => {
            let translations_tokens = translations.to_token_stream();

            quote! {
                #[doc(hidden)]
                let path: Vec<_> = #path;

                #translations_tokens
                    .find_path(&path)
                    .ok_or_else(|| translatable::Error::PathNotFound(path.join("::")))?
            }
        },
    };

    let fallback_language = config
        .fallback_language()
        .map(|lang| quote! { .or_else(|| translation_object.get(&#lang)) })
        .unwrap_or_else(|| TokenStream2::new());

    quote! {
        (|| -> Result<String, translatable::Error> {
            std::result::Result::Ok({
                #[doc(hidden)]
                let language = #language;

                #[doc(hidden)]
                let translation_object = #translation_object;
                translation_object
                    .get(&language)
                    #fallback_language
                    .ok_or_else(|| translatable::Error::LanguageNotAvailable(language, path.join("::")))?
 //                   .replace_with(&#template_replacements)
            })
        })()
    }
}
