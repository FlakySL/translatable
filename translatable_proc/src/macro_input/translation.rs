//! [`translation!()`] output generation module.
//!
//! This module declares a structure that implements
//! [`Parse`] for it to be used with [`parse_macro_input`]
//!
//! [`translation!()`]: crate::translation
//! [`parse_macro_input`]: syn::parse_macro_input

use std::collections::HashMap;

use proc_macro2::TokenStream as TokenStream2;
use quote::ToTokens;
use syn::parse::{Parse, ParseStream};
use syn::token::Static;
use syn::{Expr, ExprLit, Ident, Lit, Path, PathArguments, Result as SynResult, Token};
use thiserror::Error;
use translatable_shared::macros::errors::IntoCompileError;
use translatable_shared::misc::language::Language;

use super::input_type::InputType;

/// Parse error for [`TranslationMacroArgs`].
///
/// Represents errors that can occur while parsing the [`translation!()`]
/// macro input. This error is only used while parsing compile-time input,
/// as runtime input is validated in runtime.
#[derive(Error, Debug)]
enum TranslationMacroArgsError {
    /// An error while parsing a compile-time String value
    /// was found.
    #[error("The literal '{0}' is an invalid ISO 639-1 string, and cannot be parsed")]
    InvalidIsoLiteral(String),

    /// Extra tokens were found while parsing a static path for
    /// the [`translation!()`] macro, specifically generic arguments.
    ///
    /// [`translation!()`]: crate::translation
    #[error("This translation path contains generic arguments, and cannot be parsed")]
    InvalidPathContainsGenerics,
}

/// [`translation!()`] macro input arguments.
///
/// This structure implements [`Parse`] to parse
/// [`translation!()`] macro arguments using
/// [`parse_macro_input`], to later be used
/// in the [`translation_macro`] function.
///
/// [`translation!()`]: crate::translation
/// [`parse_macro_input`]: syn::parse_macro_input
/// [`translation_macro`]: crate::macro_generation::translation::translation_macro
pub struct TranslationMacroArgs {
    /// Represents the user specified language
    /// which may be static if the specified language
    /// is a string literal or a `Language` enum tagged
    /// union instance, otherwise dynamic and represented
    /// as a `TokenStream`.
    language: InputType<Language>,

    /// Represents a toml path to find the translation
    /// object in the previously parsed TOML from the
    /// translation files, this can be static if specified
    /// as `static path::to::translation` or dynamic if
    /// it's another expression, this way represented as a
    /// [`TokenStream2`].
    path: InputType<Vec<String>>,

    /// Stores the replacement arguments for the translation
    /// templates such as `Hello {name}` if found on a translation.
    ///
    /// If a call such as `a` is found, it will be implicitly
    /// converted to `a = a` thus stored like so in the hash map.
    replacements: HashMap<Ident, TokenStream2>,
}

/// [`translation!()`] macro args parsing implementation.
///
/// This implementation's purpose is to parse [`TokenStream`]
/// with the [`parse_macro_input`] macro.
impl Parse for TranslationMacroArgs {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let parsed_language_arg = match input.parse::<Expr>()? {
            Expr::Lit(ExprLit { lit: Lit::Str(literal), .. }) => {
                match literal
                    .value()
                    .parse::<Language>()
                {
                    Ok(language) => InputType::Static(language),

                    Err(_) => Err(TranslationMacroArgsError::InvalidIsoLiteral(literal.value())
                        .into_syn_error(literal))?,
                }
            },

            other => InputType::Dynamic(other.into_token_stream()),
        };

        input.parse::<Token![,]>()?;

        let parsed_path_arg = match input.parse::<Static>() {
            Ok(_) => {
                let language_arg = input
                    .parse::<Path>()?
                    .segments
                    .into_iter()
                    .map(|segment| match segment.arguments {
                        PathArguments::None => Ok(segment
                            .ident
                            .to_string()),

                        other => Err(TranslationMacroArgsError::InvalidPathContainsGenerics
                            .into_syn_error(other)),
                    })
                    .collect::<Result<Vec<_>, _>>()?;

                InputType::Static(language_arg)
            },

            Err(_) => InputType::Dynamic(
                input
                    .parse::<Expr>()?
                    .to_token_stream(),
            ),
        };

        let mut replacements = HashMap::new();
        if input.peek(Token![,]) {
            while !input.is_empty() {
                input.parse::<Token![,]>()?;
                let key = input.parse::<Ident>()?;
                let value = match input.parse::<Token![=]>() {
                    Ok(_) => input
                        .parse::<Expr>()?
                        .to_token_stream(),

                    Err(_) => key
                        .clone()
                        .into_token_stream(),
                };

                replacements.insert(key, value);
            }
        }

        Ok(Self {
            language: parsed_language_arg,
            path: parsed_path_arg,
            replacements,
        })
    }
}

impl TranslationMacroArgs {
    /// `self.language` reference getter.
    ///
    /// **Returns**
    /// A reference to `self.language` as [`InputType<Language>`].
    #[inline]
    #[allow(unused)]
    pub fn language(&self) -> &InputType<Language> {
        &self.language
    }

    /// `self.path` reference getter.
    ///
    /// **Returns**
    /// A reference to `self.path` as [`InputType<Vec<String>>`]
    #[inline]
    #[allow(unused)]
    pub fn path(&self) -> &InputType<Vec<String>> {
        &self.path
    }

    /// `self.replacements` reference getter.
    ///
    /// **Returns**
    /// A reference to `self.replacements` as [`HashMap<Ident, TokenStream2>`]
    #[inline]
    #[allow(unused)]
    pub fn replacements(&self) -> &HashMap<Ident, TokenStream2> {
        &self.replacements
    }
}
