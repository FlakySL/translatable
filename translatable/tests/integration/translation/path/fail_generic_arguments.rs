#[allow(unused_imports)]
use translatable::translation;

fn main() {
    translation!("es", static greetings::formal<String>);
}
