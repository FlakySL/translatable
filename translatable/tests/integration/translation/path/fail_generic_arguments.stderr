error: A translation path can't contain generic arguments.
 --> tests/integration/translation/path/fail_generic_arguments.rs:5:48
  |
5 |     translation!("es", static greetings::formal<String>);
  |                                                ^^^^^^^^
